File: LeetCode/230/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  TreeNode *ans = nullptr;

  void travel(TreeNode *node, int &k) {
    if (node == nullptr) {
      return;
    }

    travel(node->left, k);

    // at the smallest node
    k -= 1;

    if (k == 0) {
      ans = node;
      return;
    }

    travel(node->right, k);

    return;
  }

  int kthSmallest(TreeNode *root, int k) {
    travel(root, k);
    return ans->val;
  }
};

File: LeetCode/230/main.go
package 
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func traverse(node *TreeNode, num *int, k int, ans *int) {

    if node == nil {
        return
    }

    traverse(node.Left, num, k, ans)

    (*num) += 1
    if (*num) == k {
        (*ans) = node.Val
        return
    }

    traverse(node.Right, num, k, ans)
}

func kthSmallest(root *TreeNode, k int) int {
    ans, num := 0, 0
    traverse(root, &num, k, &ans)
    return ans
}